File: src/data/copilot/First_Missing_Positive/First_Missing_Positive.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        
        // Step 1: Place each number in its right place
        for (int i = 0; i < n; i++) {
            while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
                int temp = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = temp;
            }
        }
        
        // Step 2: Find the first place where the number is not correct
        for (int i = 0; i < n; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        
        return n + 1;
    }
}

public class First_Missing_Positive {
public static void runTests(Solution solution) {
    // Test case 1: Positive integers starting from 1, with no missing numbers
    int[] nums1 = {1, 2, 3};
    int expected_result1 = 4;
    assert solution.firstMissingPositive(nums1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Unsorted array with a missing positive integer in the middle
    int[] nums2 = {3, 4, -1, 1};
    int expected_result2 = 2;
    assert solution.firstMissingPositive(nums2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: No positive integers, the smallest missing positive should be 1
    int[] nums3 = {7, 8, 9, 11, 12};
    int expected_result3 = 1;
    assert solution.firstMissingPositive(nums3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Array with a mix of positive, negative, and zero
    int[] nums4 = {1, 2, 0};
    int expected_result4 = 3;
    assert solution.firstMissingPositive(nums4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Array with all negative numbers
    int[] nums5 = {-1, -2, -3};
    int expected_result5 = 1;
    assert solution.firstMissingPositive(nums5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Array with duplicates and missing a middle positive integer
    int[] nums6 = {7, 8, 9, 11, 12};
    int expected_result6 = 1;
    assert solution.firstMissingPositive(nums6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Empty array, should return 1
    int[] nums7 = {};
    int expected_result7 = 1;
    assert solution.firstMissingPositive(nums7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}


    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 2000000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/4o/First_Missing_Positive/First_Missing_Positive.cpp
#include <vector>
#include <algorithm>
#include <functional>
using namespace std;

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();

        // Step 1: Place each number in its correct position if possible
        for (int i = 0; i < n; i++) {
            while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
                swap(nums[i], nums[nums[i] - 1]);
            }
        }

        // Step 2: Find the first number that is not in its correct position
        for (int i = 0; i < n; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }

        // If all numbers are in their correct positions, return n + 1
        return n + 1;
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution &solution)
{
    // Test case 1: Positive integers starting from 1, with no missing numbers
    std::vector<int> nums1 = {1, 2, 3};
    int expected_result = 4;
    assert(solution.firstMissingPositive(nums1.data(), nums1.size()) == expected_result && "Test case 1 failed.");

    // Test case 2: Unsorted array with a missing positive integer in the middle
    std::vector<int> nums2 = {3, 4, -1, 1};
    expected_result = 2;
    assert(solution.firstMissingPositive(nums2.data(), nums2.size()) == expected_result && "Test case 2 failed.");

    // Test case 3: No positive integers, the smallest missing positive should be 1
    std::vector<int> nums3 = {7, 8, 9, 11, 12};
    expected_result = 1;
    assert(solution.firstMissingPositive(nums3.data(), nums3.size()) == expected_result && "Test case 3 failed.");

    // Test case 4: Array with a mix of positive, negative, and zero
    std::vector<int> nums4 = {1, 2, 0};
    expected_result = 3;
    assert(solution.firstMissingPositive(nums4.data(), nums4.size()) == expected_result && "Test case 4 failed.");

    // Test case 5: Array with all negative numbers
    std::vector<int> nums5 = {-1, -2, -3};
    expected_result = 1;
    assert(solution.firstMissingPositive(nums5.data(), nums5.size()) == expected_result && "Test case 5 failed.");

    // Test case 6: Array with duplicates and missing a middle positive integer
    std::vector<int> nums6 = {1, 1, 2, 3};
    expected_result = 4;
    assert(solution.firstMissingPositive(nums6.data(), nums6.size()) == expected_result && "Test case 6 failed.");

    // Test case 7: Empty array, should return 1
    std::vector<int> nums7 = {};
    expected_result = 1;
    assert(solution.firstMissingPositive(nums7.data(), nums7.size()) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 5242880; ++i) {
        testSolution(solution);  
    }
    return 0;
}
